File: app/src/main/java/com/bsoft/inventorymanager/repositories/ProductRepository.java
package com.bsoft.inventorymanager.repositories;

import androidx.lifecycle.LiveData;
import com.bsoft.inventorymanager.models.Product;
import com.google.firebase.firestore.DocumentSnapshot;
import java.util.List;

public interface ProductRepository {
    interface PaginationCallback {
        void onSuccess(List<DocumentSnapshot> documents, boolean hasMore);

        void onError(Exception e);
    }

    void fetchPaginatedProducts(
            DocumentSnapshot lastVisible,
            String brand,
            String category,
            String searchQuery,
            int pageSize,
            PaginationCallback callback);

    void fetchUniqueBrandsAndCategories(UniqueFieldsCallback callback);

    interface UniqueFieldsCallback {
        void onSuccess(List<String> brands, List<String> categories);

        void onError(Exception e);
    }

    void saveProduct(Product product, ProductCallback callback);

    void deleteProduct(Product product, ProductCallback callback);

    interface ProductCallback {
        void onSuccess();

        void onError(Exception e);
    }
}

File: shared/src/commonMain/kotlin/com/bsoft/inventorymanager/repository/ProductRepository.kt
package com.bsoft.inventorymanager.repository

import com.bsoft.inventorymanager.model.Product

interface ProductRepository {
    /**
     * Fetch paginated products with optional filters.
     * @param lastDocumentId ID of the last document for cursor-based pagination, null for first page
     * @param brand Optional brand filter
     * @param category Optional category filter
     * @param searchQuery Optional search query
     * @param pageSize Number of products per page
     * @return Result with list of products and whether more pages exist
     */
    suspend fun fetchPaginatedProducts(
        lastDocumentId: String?,
        brand: String?,
        category: String?,
        searchQuery: String?,
        pageSize: Int
    ): Result<ProductPage>
    
    suspend fun fetchUniqueBrandsAndCategories(): Result<BrandsAndCategories>
    
    suspend fun saveProduct(product: Product): Result<Unit>
    
    suspend fun deleteProduct(product: Product): Result<Unit>
    
    suspend fun getProduct(productId: String): Result<Product>
}

data class ProductPage(
    val products: List<Product>,
    val hasMore: Boolean,
    val lastDocumentId: String?
)

data class BrandsAndCategories(
    val brands: List<String>,
    val categories: List<String>
)
